### PR TITLE
[PS-1306] Context Menu for MV3

### PR DESCRIPTION
**Pull request by justindbaur**
## Type of change

```
- [x] Bug fix
- [x] New feature development
- [x] Tech debt (refactoring, code cleanup, dependency upgrades, etc)
- [ ] Build/deploy pipeline (DevOps)
- [ ] Other
```

## Objective

This PR adds context menu functionality to MV3 extensions through the use of service workers. It also builds it in such a way that it is able to scaffold itself for the service worker contexts but is able to utilize service caching of having a background page. It is able to be the home of all the context menu logic and replace the logic spread between `main.background.ts` and `contextMenus.background.ts`. 

## Code changes

- **apps/browser/src/background.ts:** Utilizes a helper method to combine multiple listeners into one listener that allows sharing of cached services.
- **apps/browser/src/background/contextMenus.background.ts:** Remove context menu clicked logic from here.
- **apps/browser/src/background/main.background.ts:** Removed most context menu logic in favor of scaffolding context menu sub-services only when not in popup only context (MV2).
- **apps/browser/src/browser/context-menu-clicked-handler.ts:** Added a service that has the job of handling clicks to context menu items. Much of this logic is from `contextMenus.background.ts` but done so in a way that is more MV3 friendly. 
- **apps/browser/src/browser/cipher-context-menu-handler.ts:** Add a service that has the job of adding cipher related data onto the contextMenu. This could be combined with `MainContextMenuHandler` but that would require more services than that service would need much of the time.
- **apps/browser/src/browser/main-context-menu-handler.ts:** This handles the scaffolding of the skeleton context menu states. Like our root object, top-level categories and then the `noAccess` or `noLogins` states.
- **apps/browser/src/listeners/update-badge.ts:** Changed the method signatures so that they would be more `combine` friendly. 
- Removed some `any`

## Before you submit

- Please add **unit tests** where it makes sense to do so (encouraged but not required)
- If this change requires a **documentation update** - notify the documentation team
- If this change has particular **deployment requirements** - notify the DevOps team

### CODE REVIEW

File: apps/browser/src/background/main.background.ts [hunk 1]
```diff
@@ -551,6 +558,14 @@ export default class MainBackground {
...
       this.apiService
     );
+
+    this.mainContextMenuHandler = new MainContextMenuHandler(this.stateService, this.i18nService);
+
+    this.cipherContextMenuHandler = new CipherContextMenuHandler(
+      this.mainContextMenuHandler,
+      this.authService,
+      this.cipherService
+    );
```
Comment (MGibson1) [inline apps/browser/src/background/main.background.ts:568]: Why are these needed in popup only contexts?

File: apps/browser/src/background/main.background.ts [hunk 2]
```diff
@@ -568,7 +583,9 @@ export default class MainBackground {
     this.twoFactorService.init();
 
     await this.tabsBackground.init();
-    await this.contextMenusBackground.init();
+    if (!this.popupOnlyContext) {
+      this.contextMenusBackground.init();
```
Comment (MGibson1) [inline apps/browser/src/background/main.background.ts:587]: Seems smart to handle the null here, since you're not always instantiating the class.

```suggestion
      this.contextMenusBackground?.init();
```

File: apps/browser/src/browser/browserApi.ts [hunk 3]
```diff
@@ -102,6 +102,15 @@ export class BrowserApi {
...
 
+  static sendTabsMessageOptions<T>(
+    tabId: number,
+    message: TabMessage,
+    options: chrome.tabs.MessageSendOptions,
+    responseCallback?: (response: T) => void
+  ) {
+    chrome.tabs.sendMessage<TabMessage, T>(tabId, message, options, responseCallback);
+  }
+
```
Comment (MGibson1) [inline apps/browser/src/browser/browserApi.ts:113]: All of these methods from line 78 to 113  _really_ seem duplicative. What do you think about combining to a single overloaded method?

File: apps/browser/src/browser/cipher-context-menu-handler.ts [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+} from "../background/service_factories/cipher-service.factory";
+import { CachedServices } from "../background/service_factories/factory-options";
+import { searchServiceFactory } from "../background/service_factories/search-service.factory";
+import { Account } from "../models/account";
+
+import { BrowserApi } from "./browserApi";
+import { MainContextMenuHandler } from "./main-context-menu-handler";
+
+const NOT_IMPLEMENTED = (..._args: unknown[]) =>
+  Promise.reject<never>("This action is not implemented inside of a service worker context.");
```
Comment (MGibson1) [inline apps/browser/src/browser/cipher-context-menu-handler.ts:28]: From what I can tell, this is not limited to MV3 or a service worker. I think it's more reasonable to pass a noop than an intentional error

File: apps/browser/src/content/contextMenuHandler.ts [hunk 5]
```diff
@@ -54,9 +54,10 @@ document.addEventListener("contextmenu", (event) => {
 });
 
 // Runs when the 'Copy Custom Field Name' context menu item is actually clicked.
-chrome.runtime.onMessage.addListener((event) => {
+chrome.runtime.onMessage.addListener((event, _sender, sendResponse) => {
   if (event.command === "getClickedElement") {
     const identifier = getClickedElementIdentifier();
+    sendResponse(identifier);
```
Comment (MGibson1) [inline apps/browser/src/content/contextMenuHandler.ts:60]: Please null-check

**Review by MGibson1 — COMMENTED** [review]
I need to review the menu handler classes, still, but I really want to get some feedback uploaded. Sorry for the horribly long review time

File: apps/browser/src/browser/browserApi.ts [hunk 6]
```diff
@@ -102,6 +102,15 @@ export class BrowserApi {
     chrome.tabs.sendMessage<TabMessage, T>(tabId, message, responseCallback);
   }
 
+  static sendTabsMessageOptions<T = never>(
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply apps/browser/src/browser/browserApi.ts]: Can't overload functions in JS unless we just take an array of args and parse it out.

[the remark this answers is not in the document]
Comment (justindbaur) [reply apps/browser/src/browser/browserApi.ts]: We had a discussion about this and it was removed.

File: apps/browser/src/browser/browserApi.ts (hunk 3 above)
Comment (justindbaur) [reply apps/browser/src/browser/browserApi.ts:113]: Actually did make it just one function with 4 arguments.

File: apps/browser/src/background/main.background.ts (hunk 1 above)
Comment (justindbaur) [reply apps/browser/src/background/main.background.ts:568]: They are not, put them behind the `popupOnlyContext` check.

File: apps/browser/src/listeners/update-badge.ts [hunk 7]
```diff
@@ -245,6 +266,9 @@ export class UpdateBadge {
       i18nServiceOptions: {
         systemLanguage: BrowserApi.getUILanguage(self),
       },
+      cipherServiceOptions: {
+        searchServiceFactory: () => searchService,
+      },
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply apps/browser/src/listeners/update-badge.ts:271]: I think this was the result of a bad merge resolution on my part. I have no clue why it allowed that to build... but I removed it now.

File: apps/browser/src/listeners/update-badge.ts [hunk 8]
```diff
@@ -220,7 +241,7 @@ export class UpdateBadge {
     }
 
     const serviceCache: Record<string, unknown> = existingServiceCache || {};
-    const opts = {
+    const opts: StateServiceInitOptions & AuthServiceInitOptions & CipherServiceInitOptions = {
```
[the remark this answers is not in the document]
Comment (justindbaur) [reply apps/browser/src/listeners/update-badge.ts]: I find the type pretty required when I am building them up but they aren't needed as much after. I remove them on this one.

**Review by djsmith85 — APPROVED** [review]
Changes are looking good.

**Review by MGibson1 — APPROVED** [review]

**Review by djsmith85 — DISMISSED** [review]

**Review by djsmith85 — DISMISSED** [review]

**Review by djsmith85 — APPROVED** [review]

**Review by djsmith85 — APPROVED** [review]
